README.md

[![Arduino CI](https://github.com/RobTillaart/TM1637_RT/workflows/Arduino%20CI/badge.svg)](https://github.com/marketplace/actions/arduino_ci)
[![Arduino-lint](https://github.com/RobTillaart/TM1637_RT/actions/workflows/arduino-lint.yml/badge.svg)](https://github.com/RobTillaart/TM1637_RT/actions/workflows/arduino-lint.yml)
[![JSON check](https://github.com/RobTillaart/TM1637_RT/actions/workflows/jsoncheck.yml/badge.svg)](https://github.com/RobTillaart/TM1637_RT/actions/workflows/jsoncheck.yml)
[![GitHub issues](https://img.shields.io/github/issues/RobTillaart/TM1637_RT.svg)](https://github.com/RobTillaart/TM1637_RT/issues)

[![License: MIT](https://img.shields.io/badge/license-MIT-green.svg)](https://github.com/RobTillaart/TM1637_RT/blob/master/LICENSE)
[![GitHub release](https://img.shields.io/github/release/RobTillaart/TM1637_RT.svg?maxAge=3600)](https://github.com/RobTillaart/TM1637_RT/releases)
[![PlatformIO Registry](https://badges.registry.platformio.org/packages/robtillaart/library/TM1637_RT.svg)](https://registry.platformio.org/libraries/robtillaart/TM1637_RT)


# TM1637

Library for TM1637 driven displays and key scans.


## Description

The TM1637 drives 7 segment displays and can also scan a 16 key keyboard.

The library is mainly tested with Arduino UNO, both a 6 digits display and 4 digit (clock) display.

ESP32 is supported since 0.2.0 see https://github.com/RobTillaart/TM1637_RT/pull/5


#### Related

- https://docs.wokwi.com/parts/wokwi-tm1637-7segment#simulator-examples
- https://github.com/SteveMicroCode/PIC-TM1637-Library-CodeOn 
  - includes interesting hardware notes.


#### Hardware connection and performance

From tests of SteveMicroCode (See #29) it became clear that the proposed hardware in the
datasheet is very robust but slow. See page 8 datasheet.
Leaving out the 100 pF capacitors on the CLK and DIO line, or dimension them smaller,
allows for faster communication and thus more efficient CPU use. 
More about his tests see - https://github.com/SteveMicroCode/PIC-TM1637-Library-CodeOn 

Additional tests of wfdudley (See Keyboard Scanner section) shows that reducing
the pull up resistor from 10 KΩ to 1000 Ω improves the signal quality.
This too allows for faster communication and efficient CPU use.

Datasheet page 3: _and the clock frequency should be less than 250K_ (KHz)   
This implies that the **setBitDelay()** could be set much lower than the default 10 micro
seconds of the library. On "slower" platforms it might even be 0.
However it is always important to do your own tests if you want to tweak performance
beyond the specifications of the datasheet. 
Interference from other electronic components is often nearby, and if you need longer
cables consider shielded ones.

If you have additional performance tweaks, please let me know.

Note: Ω = ALT-234


## Interface

```cpp
#include "TM1637.h"
```

#### Core

- **TM1637()** constructor
- **void begin(uint8_t clockPin, uint8_t dataPin, uint8_t digits = 6)**
 set up the connection of the pins to the display.
As the display was tested with a 6 digit display first, 
this is used as the default of the digits parameter.


#### Display functions I

- **void displayClear()** writes spaces to all positions, effectively clearing the display.
- **void displayRefresh()** refreshes last written data on display.
To be used e.g. after a **hideSegment()** call.
- **void hideSegment(uint8_t idx)** writes space to a single segment.
- **void hideMultiSegment(uint8_t mask)** writes spaces to 0 or more segments depending on the mask.


#### Display functions II

- **void displayPChar(char \*buff)** display the buffer. 
Experimental - Tested on STM32 and Arduino Nano.
- **void displayRaw(uint8_t \* data, uint8_t pointPos)** low level write function.
Display a point at pointPos (see below).
- **void displayInt(long value)** displays an integer value.
- **void displayFloat(float value)** displays a float value.
Position of point may vary!
- **void displayFloat(float value, uint8_t fixedPoint)** displays a float value with a fixed point position.
- **void displayHex(uint32_t value)** display a number in hexadecimal notation.
- **void displayTime(uint8_t hh, uint8_t mm, bool colon)** displays time format.
The function does not check for overflow e.g. hh > 59 or mm > 59.
Works only on a 4 digit display.
Can be used for 
  - hours + minutes HH:MM 
  - minutes + seconds MM:SS
  - seconds + hundreds SS:hh
The colon can be used e.g. as pulses to indicate seconds etc.
- **void displayTwoInt(int ll, int rr, bool colon = true)** print two integers,
 one left and one right of the colon. 
The function allows a range from -9 .. 99 (not checked).
The colon is default on as separator.
Works only on a 4 digit display.
Applications include:
  - temperature + humidity TT:HH  (humidity to 99%)
  - heartbeat + oxygen HH:OO (both max to 99)
  - meters + centimetre MM:CC  (e.g distance sensor)
  - feet + inches FF:II
  - any pair of integers (-9 .. 99) side by side.


#### Display functions III

- **void displayCelsius(int temp, bool colon = false)** print temperature in **Celsius** format.
The function allows a range from -9 .. 99 + °C.
The colon is default false.
Works only on a 4 digit display.
Colon can be used e.g. to indicate under- or overflow, or any other threshold.
- **void displayFahrenheit(int temp, bool colon = false)** print temperature in **Fahrenheit** format.
The function allows a range from -9 .. 99 + °F.
The colon is default false.
Works only on a 4 digit display.
Colon can be used e.g. to indicate under- or overflow, or any other threshold.

```cpp
TM.displayCelsius(temperature, (temperature < -9) || (temperature > 99));
```

Note that the effective range of Celsius and Fahrenheit differs.
When Fahrenheit goes from -9 to 99 Celsius goes from -26 to 37 (etc).

|   F   |   C   |      |   F   |   C   |
|:-----:|:-----:|:----:|:-----:|:-----:|
|  -9   | -26   |      |   16  |  -9   |
|  99   |  37   |      |  210  |  99   |

A three digit temperature, e.g. range -99 .. 999, would be possible but one has 
to leave out either the ° character or the C/F.
As the C/F is more informative, the choice is fairly easy.
Question is how to API should change, new function of new behaviour?
See future.


#### Brightness

- **void setBrightness(uint8_t brightness = 3)** brightness = 0 .. 7 default = 3.
- **uint8_t getBrightness()** returns value set.


#### KeyScan

- **uint8_t keyscan(void)** scans the keyboard once and return result. 
The keyscan() function cannot detect multiple keys.


#### DisplayRaw explained

**displayRaw()** can display multiple decimal points, by setting the high bit (0x80) 
in each character for which you wish to have a decimal point lit.  
Or you can use the pointPos argument to light just one decimal at that position.

**displayRaw()** can display some of the alphabet as follows:
   - space (blank) is 0x10
   - '-' (minus) is 0x11
   - a-f are coded as 0x0a-0x0f
   - g-z are coded as 0x12-0x25.  Characters that cannot be represented in 7 segments render as blank.
So "hello " is coded as 0x13, 0x0e, 0x17, 0x17, 0x1a, 0x10


#### displayPChar explained

**void displayPChar(char \* buff)** Attempts to display every ASCII character 0x30 to 0x5F. 
See example TM1637_custom.ino to insert your own 7 segment patterns.
Also displayed are  '  ' , '.' and '-' . Decimal points may also be displayed by setting the character sign bit.

See routine **ascii_to_7segment()** in the example TM1637_keyscan_cooked.ino. 
It presents a more convenient interface for displaying text messages on the display.

Routine **button_poll()** in the same example shows one way of polling and de-bouncing button presses.


#### hideSegment() explained

- **void hideSegment(uint8_t idx)** hides a single segment until any other call to display.
- **void hideMultiSegment(uint8_t mask)** hides 0 or more segments depending on the mask.
A 0 bit shows the segment, and a 1 bit hides the segment. 
A mask of 0x00 == 0b00000000 will show all segments.
A mask 0f 0xFF == 0b11111111 will hide all segments. 
- **void displayRefresh()** refreshes last written data on display.

The function **hideSegment()** can be used to
- set focus on an digit, e.g. during an input
- to create an animation e.g. "wave"

The function **hideMultiSegment()** can be used to
- hide a partial display, e.g. in combination with the **displayTwoInt()** 
to alternate between the two values.
- put a display in a LOW ENERGY mode. **displayRefresh()** will refresh the last written data.
- put a display in a DARK mode, e.g. only show the data when a key is pressed (any trigger).
- create an animation, 
- implement blinking in combination with **displayRefresh()**.

Note: **hideMultiSegment()** and **hideMultiSegment()** do not affect the "display cache" 
where **displayClear()** fills the "display cache" with spaces.


#### Obsolete (0.4.0)

- **void init(uint8_t clockPin, uint8_t dataPin, uint8_t digits = 6)** replaced by begin().


#### Display support

The library is tested with a 6 (=2x3) digit - decimal point - display and a 4 (=1x4) digit - clock - display. 
At low level these displays differ in the order the digits have to be clocked in.
To adjust the order for a not supported display, the following function can be used with care:

- **void setDigitOrder(uint8_t a, uint8_t b,... uint8_t h)** sets the order in which up to 8 digits should be clocked in.

If you have a (7 segment) display that is not supported by the library, 
please open an issue on GitHub so it can be build in.


#### Tuning function

**setBitDelay()** is used to tune the timing of writing bytes. 
An UNO can gain up to 100 micros per call by setting the bit delay to 0.
However some displays might fail with short bit delay's.
Do not forget to use a pull up resistor on the clock and data line.

- **void setBitDelay(uint8_t bitDelay = 10)**
- **uint8_t getBitDelay()**


#### Tuning minimum pulse length

The class has a conditional code part in writeSync to guarantee the length of pulses
when the library is used with an ESP32. The function called there **nanoDelay(n)**
needs manual adjustment depending upon processor frequency and time needed for a digitalWrite.
Feel free to file an issue to get your processor supported.

----

## Keyboard Scanner usage and notes

- Kudos to wfdudley for this section - See #11

Calling **keyscan()** returns a uint8_t, whose value is 0xff if no keys are being pressed at the time.  
The TM1637 can only see one key press at a time, and there is no "rollover".  
If a key is pressed, then the values are as follows:

<CENTER>
<TABLE>
<TR>
<TD colspan = 10 align="center">
   keyscan results are reversed left for right from the data sheet.
</TD>
</TR>
<TR>
<TH>pin</TH><TD>&nbsp</TD><TD>2</TD><TD>3</TD><TD>4</TD><TD>5</TD><TD>6</TD><TD>7</TD><TD>8</TD><TD>9</TD>
</TR>
<TR>
</TD><TD>&nbsp;</TD><TH>name</TH><TD>sg1</TD><TD>sg2</TD><TD>sg3</TD><TD>sg4</TD><TD>sg5</TD><TD>sg6</TD><TD>sg7</TD><TD>sg8</TD>
</TR>
<TR>
</TD><TD>19</TD><TD>k1</TD><TD>0xf7</TD><TD>0xf6</TD><TD>0xf5</TD><TD>0xf4</TD><TD>0xf3</TD><TD>0xf2</TD><TD>0xf1</TD><TD>0xf0</TD>
</TR>
<TR>
</TD><TD>20</TD><TD>k2</TD><TD>0xef</TD><TD>0xee</TD><TD>0xed</TD><TD>0xec</TD><TD>0xeb</TD><TD>0xea</TD><TD>0xe9</TD><TD>0xe8</TD>
</TR>
</TABLE>
</CENTER>


To modify a "generic" TM1637 board for use with a keyboard, you must add connections to either
or both of pins 19 and 20 (these are the "row" selects) and then to as many of pins 2 through 9
(the "columns") as needed.  It is easiest to connect to the "column pins" (2-9) by picking them
up where they connect to the LED displays (see second photo).
Generic keyboards that are a 4x4 matrix won't work; the TM1637 can only scan a 2x8 matrix.
Of course, fewer keys are acceptable; I use a 1x4 keyboard in my projects.

Further, the TM1637 chip needs a fairly hefty pull-up on the DIO pin for the keyscan() routine to work.
There is no pull-up in the TM1637 itself, and the clone boards don't seem to have one either,
despite the data sheet calling for 10K ohms pull-ups on DIO and CLOCK.  10K is too weak anyway.
The slow rise-time of the DIO signal means that the "true/high" value isn't reached fast enough and 
reliably enough for the processor to read it correctly. The new pull-up reduces the rise time of the signal, 
so that true/high values are achieved in a few microseconds.
I find that a 1K (1000) ohm resistor from DIO to 3.3 v works well. This is perfect with a 3.3 volt processor 
like the ESP8266 or ESP32, and a 5V Atmega 328 ("Arduino UNO") family processor is happy with that as well.

The TM1637 boards want to be run off of 5 volts, regardless of what the processor voltage is.
Their logic levels are compatible with 3.3 volt processors, and they need 5 volts to make sure the LEDs light up.


The unmodified generic TM1637 board (front and back).

<IMG src="images/unmodified.jpg">


The modified generic TM1637 board with connector for 1x4 keyboard. 
The blue wire is bringing out pin 19 (k1).
Four segments/columns are picked up from the LEDs.

<IMG src="images/modified.jpg">

The 4 button keyboard plugged into the TM1637 board.

<IMG src="images/disp_plus_kbd.jpg">


Scope photo showing slow rise time of DIO pin (upper trace) on the unmodified TM1637.
The lower trace is the CLK.  The 8 fast CLK pulses on the left represent the 0x42 command 
to read keyboard being sent to the TM1637.

<IMG src="images/slow_rise.jpg">


Scope photo showing faster rise time of DIO pin (upper trace) with 1000 ohm pull-up on DIO.
In both scope photos, the F5 key is pressed; the bits are least significant bit (LSB) first,
so read as 10101111 left to right.

<IMG src="images/fast_rise.jpg">

The scope photos were taken using the TM1637_keyscan_raw example, with the scope trigger 
hooked to the TRIGGER pin, and the two channel probes hooked to DIO and CLK.
Vertical sensitivity is 2v/division, horizontal timebase is 20usec/division.


## Keyscan

Implemented in version 0.3.0  Please read the datasheet to understand the limitations.

```
//  NOTE: 
//  on the TM1637 boards tested by @wfdudley, keyscan() works well 
//  if you add a 910 ohm or 1 Kohm pull-up resistor from DIO to 3.3v
//  This reduces the rise time of the DIO signal when reading the key info.
//  If one only uses the pull-up inside the microcontroller, 
//  the rise time is too long for the data to be read reliably.
```


## Operation

See examples


## Future

#### Must

- (0.4.0)
  - remove obsolete **init()** from code
  - **setLeadingZeros(bool on = false)** leading zeros flag, set data array to 0.
  - **getLeadingZeros()**

#### Should

- testing other platforms.
- refactor readme.md
- remove degree sign from **displayCelsius()**
  - would allow one extra digit.
  - **displayFahrenheit()** idem.
  - could be optional when needed e.g. below -9 or above 99

#### Could

- **keyScan()** camelCase ?
- add parameter for **hideSegement(idx, character == SPACE)** to overrule hide char.
  - space underscore or - are possible.
- add **TM1637_UNDERSCORE** to char set. ```seg[19] == 0x08```
- add **TM1637_UPPERSCORE** to char set. ```seg[20] == 0x01```
- return bytes written as return value for display functions.
  - like print(); or write();
- **void displayHumidity(int hum, bool colon = false)** print humidity in **percent** format.
The function allows a range from 0.00H .. 99.9H or 00.0H ?
Colon to indicate rising / falling ????
- generic function that displays three digits and one char ACDEFH ???
  - common for Celsius and Fahrenheit and Humidity?
- **void displayAmpere(float amps)** Ampere 0.00 - 999 A


#### Wont (unless requested)

- **rotate(bool rot = false)**
  - 180 degree rotation of all digits for mounting  
  - reverse digit order
  - flip every digit (function to overwrite the char array)
- **HUD(bool hud = false)** = Heads Up Display
  - flip every digit
- investigate if code can be optimized 
  - done, => tune setBitDelay()
  - performance measurement
- **displayTest()** function ?
  - not needed just print 88888888
- add debug flag for test without hardware.
  - simulate output to Serial? (HEX)?
  - is now commented code, good enough + dumpcache()

CHANGELOG.md
# Change Log TM1637

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).


## [0.3.9] - 2023-11-22
- update readme.md
  - section about hardware performance (Kudos to SteveMicroCode #29)
  - incl badges to platformIO


## [0.3.8] - 2023-07-15
- fix #27
- replaced function data\[8] with a class level data\[8].
- add **hideSegment(uint8_t n)** hide a single segment
- add **hideMultiSegment(uint8_t mask)** hide multiple segments
- add **displayFahrenheit()** -9..99F
- add **displayRefresh()** to refresh last written data.
- update keywords.txt
- fix bug in **displayRaw()** (add bool hasPoint).
- update readme.md

## [0.3.7] - 2023-04-17
- move code from .h to .cpp
- add **displayTwoInt(int ll, int rr, bool colon = true)**
- add **displayCelsius(int temp, bool colon = false)**
- add '°' degree character (18)
- add examples
- update readme.md

## [0.3.6] - 2023-02-27
- add **void displayTime(uint8_t hh, uint8_t mm, bool colon)**
- add examples
- update readme.md
- update keywords.txt

## [0.3.5] - 2023-02-18
- add **void displayFloat(float value, byte fixpoint)** Thanks to marshalab
- add example TM1637_float_point.ino
- update readme.md
- update GitHub actions
- update license 2023
- minor edits

## [0.3.4] - 2022-10-07
- added CHANGELOG.md
- added **void displayPChar(char \* data)** thanks to radionerd
- added **uint8_t asciiTo7Segment (char c)** thanks to radionerd
- updated documentation

## [0.3.3] - 2022-09-24
- added CHANGELOG.md
- added **begin()** to replace **init()** in due time.
- updated documentation

## [0.3.2] - 2022-04-16
- fix #15 support for 4 digits.
- tested on 4 digit (clock) display.

## [0.3.1] - 2021-12-29
- update library.json
- update license
- update readme.md
- minor edits
- tested on 6 digits display (with decinal point).

## [0.3.0] - 2021-10-27
- improved keyscan - kudos to wfdudley
- update documentation

----

## [0.2.1] - 2021-10-07
- add support for letters g-z; added keyscan()
- tested on ESP8266

## [0.2.0] - 2021-09-26
- add ESP32 support - kudos to alexthomazo

----

## [0.1.2] - 2021-04-16
- update readme
- fix default values

## [0.1.1] - 2021-02-15
- first release
- added examples

## [0.1.0] - 2019-10-28
- initial version



TM1637.cpp
//
//    FILE: TM1637.cpp
//  AUTHOR: Rob Tillaart
//    DATE: 2019-10-28
// VERSION: 0.3.9
// PURPOSE: TM1637 library for Arduino
//     URL: https://github.com/RobTillaart/TM1637_RT

//  NOTE:
//  on the inexpensive TM1637 boards @wfdudley has used, keyscan
//  works if you add a 1000 ohm pull-up resistor from DIO to 3.3v
//  This reduces the rise time of the DIO signal when reading the key info.
//  If one only uses the pull-up inside the microcontroller,
//  the rise time is too long for the data to be read reliably.


#include "TM1637.h"


#define TM1637_ADDR_AUTO           0x40
#define TM1637_READ_KEYSCAN        0x42
#define TM1637_ADDR_FIXED          0x44

#define TM1637_CMD_SET_DATA        0x40
#define TM1637_CMD_SET_ADDR        0xC0
#define TM1637_CMD_DISPLAY         0x88

//  Special chars
#define TM1637_SPACE      16
#define TM1637_MINUS      17
#define TM1637_DEGREE     18


/***************
   ---
  |   |
   ---
  |   |
   ---    .


      -01-
  20 |    | 02
      -40-
  10 |    | 04
      -08-    .80

*/


// PROGMEM ?


static uint8_t seg[] =
{
  0x3f, 0x06, 0x5b, 0x4f, 0x66, 0x6d, 0x7d, 0x07, 0x7f, 0x6f,   // 0 - 9
  0x77, 0x7c, 0x39, 0x5e, 0x79, 0x71, 0x00, 0x40, 0x63          // A - F, ' ', '-', '°'
};


static uint8_t alpha_seg[] =
{
    0x00, 0x74, 0x10, 0x00,      // g, h, i, j,
    0x00, 0x38, 0x00, 0x54,      // k, l, m, n,
    0x5c, 0x00, 0x00, 0x50,      // o, p, q, r,
    0x00, 0x31, 0x1c, 0x1c,      // s, t, u, v,
    0x00, 0x00, 0x00, 0x00       // w, x, y, z
};


TM1637::TM1637()
{
  _digits     = 6;
  _brightness = 3;
  _bitDelay   = 10;
}


//  wrapper, init() will become obsolete 0.4.0.
void TM1637::init(uint8_t clockPin, uint8_t dataPin, uint8_t digits)
{
  begin(clockPin, dataPin, digits);
}


void TM1637::begin(uint8_t clockPin, uint8_t dataPin, uint8_t digits)
{
  _clockPin = clockPin;
  _dataPin  = dataPin;
  _digits   = digits;

  pinMode(_clockPin, OUTPUT);
  digitalWrite(_clockPin, HIGH);
  pinMode(_dataPin, OUTPUT);
  digitalWrite(_dataPin, HIGH);

  //  TODO: replace _digits by a display enumeration?
  if (_digits == 4)
  {
    setDigitOrder(3, 2, 1, 0);
  }
  else  //  (_digits == 6 )    //  default
  {
    setDigitOrder(3, 4, 5, 0, 1, 2);
  }
}


void TM1637::displayInt(long value)
{
  for (int i = 0; i < 8; i++) _data[i] = TM1637_SPACE;  //  16

  long v = value;
  int last = _digits;
  bool neg = (v < 0);
  if (neg)
  {
    v = -v;
    last--;
    _data[last] = TM1637_MINUS;
  }

  for (int i = 0; i < last; i++)
  {
    long t = v / 10;
    _data[i] = v - 10 * t;   //  faster than %
    v = t;
  }

  displayRaw(_data, -1);
}


void TM1637::displayFloat(float value)
{
  for (int i = 0; i < 8; i++) _data[i] = TM1637_SPACE;  //  16

  float v  = value;
  int dpos = _digits - 1;
  int last = _digits;
  bool neg = (v < 0);

  if (neg)
  {
    v = -v;
    dpos--;
    last--;
    _data[last] = TM1637_MINUS;
  }

  while (v >= 10)
  {
    v /= 10;
    dpos--;
  }
  for (int i = last-1; i > -1; i--)
  {
    int d = v;
    _data[i] = d;
    v -= d;
    v *= 10;
  }
  displayRaw(_data, dpos);
}


void TM1637::displayFloat(float value, uint8_t fixedPoint)
{
  for (int i = 0; i < 8; i++) _data[i] = TM1637_SPACE;  //  16

  float v   = value;
  int dpos  = _digits - 1;
  int last  = _digits;
  bool neg  = (v < 0);
  int point = fixedPoint + 1;

  if (neg)
  {
    v = -v;
    dpos--;
    last--;
  }
  //  v += 0.0001; //  Bug fix for 12.999 <> 13.000
  v += 0.001;      //  Bug fix for 12.99 <> 13.00

  while (v >= 10)
  {
    v /= 10;
    dpos--;
    point++;
  }

  if (neg)
  {
    _data[point] = TM1637_MINUS;
  }

  for (int i = point - 1; i > -1; i--)
  {
    int d = v;
    _data[i] = d;
    v -= d;
    v *= 10;
  }
  displayRaw(_data, fixedPoint);
}


void TM1637::displayHex(uint32_t value)
{
  for (int i = 0; i < 8; i++) _data[i] = TM1637_SPACE;  //  16

  uint32_t v = value;
  for (int i = 0; i < _digits; i++)
  {
    uint32_t t = v / 16;
    _data[i] = v & 0x0F;     //  faster than %
    v = t;
  }
  displayRaw(_data, -1);
}


void TM1637::displayTime(uint8_t hour, uint8_t minute, bool colon)
{
  if (_digits != 4) return;
  for (int i = 0; i < 8; i++) _data[i] = TM1637_SPACE;  //  16

  //  optional
  //  if (hour > 99) hour = 99;
  //  if (minute > 99) minute = 99;
  _data[3] = hour / 10;
  _data[2] = hour % 10;
  _data[1] = minute / 10;
  _data[0] = minute % 10;
  displayRaw(_data, colon ? 2 : -1);
}


void TM1637::displayTwoInt(int left, int right, bool colon)
{
  if (_digits != 4) return;
  for (int i = 0; i < 8; i++) _data[i] = TM1637_SPACE;  //  16

  //  optional
  //  if (left < -9)  left = -9;
  //  if (left > 99)  left = 99;
  //  if (right < -9) right = -9;
  //  if (right > 99) right = 99;
  if (left < 0)
  {
    _data[3] = TM1637_MINUS;
    _data[2] = -left;
  }
  else
  {
    _data[3] = left / 10;
    _data[2] = left % 10;
  }
  if (right < 0)
  {
    _data[1] = TM1637_MINUS;
    _data[0] = -right;
  }
  else
  {
    _data[1] = right / 10;
    _data[0] = right % 10;
  }
  displayRaw(_data, colon ? 2 : -1);
}


void TM1637::displayCelsius(int temp, bool colon)
{
  if (_digits != 4) return;
  for (int i = 0; i < 8; i++) _data[i] = TM1637_SPACE;  //  16
  _data[0] = 12;             //  C
  _data[1] = TM1637_DEGREE;  //  ° degree sign

  if (temp < -9) temp = -9;
  if (temp > 99) temp = 99;
  if (temp < 0)
  {
    _data[3] = TM1637_MINUS;
    _data[2] = -temp;
  }
  else
  {
    _data[3] = temp / 10;
    _data[2] = temp % 10;
  }
  displayRaw(_data, colon ? 2 : -1);
}


void TM1637::displayFahrenheit(int temp, bool colon)
{
  if (_digits != 4) return;
  for (int i = 0; i < 8; i++) _data[i] = TM1637_SPACE;  //  16
  _data[0] = 15;             //  F
  _data[1] = TM1637_DEGREE;  //  ° degree sign

  if (temp < -9) temp = -9;
  if (temp > 99) temp = 99;
  if (temp < 0)
  {
    _data[3] = TM1637_MINUS;
    _data[2] = -temp;
  }
  else
  {
    _data[3] = temp / 10;
    _data[2] = temp % 10;
  }
  displayRaw(_data, colon ? 2 : -1);
}


void TM1637::displayClear()
{
  for (int i = 0; i < 8; i++) _data[i] = TM1637_SPACE;  //  16
  displayRaw(_data, -1);
}


void TM1637::displayRefresh()
{
  //  display internal buffer again.
  displayRaw(_data, _lastPointPos);
}


void TM1637::hideSegment(uint8_t idx)
{
  if (idx > 7) return;
  uint8_t tmp[8];
  for (int i = 0; i < 8; i++) tmp[i] = _data[i];
  tmp[idx] = TM1637_SPACE;
  displayRaw(tmp, _lastPointPos);
}


void TM1637::hideMultiSegment(uint8_t mask)
{
  uint8_t tmp[8];
  for (int i = 0; i < 8; i++)
  {
    if ((mask & 0x01) == 0x01) tmp[i] = TM1637_SPACE;
    else tmp[i] = _data[i];
    mask >>= 1;
  }
  displayRaw(tmp, _lastPointPos);
}


void TM1637::setBrightness(uint8_t brightness)
{
  _brightness = brightness;
  if (_brightness > 0x07) _brightness = 0x07;
}


uint8_t TM1637::getBrightness()
{
  return _brightness;
}


void TM1637::setBitDelay(uint8_t bitDelay)
{
  _bitDelay = bitDelay;
}


uint8_t TM1637::getBitDelay()
{
  return _bitDelay;
}


void TM1637::setDigitOrder(uint8_t a, uint8_t b,
                uint8_t c, uint8_t d, uint8_t e,
                uint8_t f, uint8_t g, uint8_t h)
{
  _digitOrder[0] = a;
  _digitOrder[1] = b;
  _digitOrder[2] = c;
  _digitOrder[3] = d;
  _digitOrder[4] = e;
  _digitOrder[5] = f;
  _digitOrder[6] = g;
  _digitOrder[7] = h;
}


//  Set sign bit on any char to display decimal point
void TM1637::displayPChar( char * data )
{
  start();
  writeByte(TM1637_ADDR_AUTO);
  stop();

  start();
  writeByte(TM1637_CMD_SET_ADDR);

  for (int d = _digits-1; d >= 0 ; d--)
  {
    uint8_t i = _digitOrder[d];
    writeByte( asciiTo7Segment(data[i]) );
  }
  stop();

  start();
  writeByte(TM1637_CMD_DISPLAY | _brightness);
  stop();
}


void TM1637::displayRaw(uint8_t * raw, uint8_t pointPos)
{
  //  DEBUG
  // for (uint8_t d = 0; d < _digits; d++)
  // {
    // uint8_t x = raw[_digits - d - 1];
    // if (x < 0x10) Serial.print('0');
    // Serial.print(x, HEX);
    // Serial.print('-');
  // }
  // Serial.println();

  uint8_t b = 0;
  _lastPointPos = pointPos;

  start();
  writeByte(TM1637_ADDR_AUTO);
  stop();

  start();
  writeByte(TM1637_CMD_SET_ADDR);

  for (uint8_t d = 0; d < _digits; d++)
  {
    uint8_t i = _digitOrder[d];
    bool hasPoint = raw[i] & 0x80;
    raw[i] &= 0x7f;
    if (raw[i] <= 18)        //  HEX DIGIT
    {
      b = seg[raw[i]];
    }
    else if (raw[i] <= 37)   //  ASCII
    {
      b = alpha_seg[raw[i] - 18];
    }
    //  do we need a decimal point
    if ((i == pointPos) || hasPoint)
    {
      b |= 0x80;
    }
    writeByte(b);
  }
  stop();

  start();
  writeByte(TM1637_CMD_DISPLAY | _brightness);
  stop();
}


void TM1637::dumpCache()
{
  for (int i = 0; i < 8; i++)
  {
    Serial.print(_data[i]);
    Serial.print(" ");
  }
  Serial.println();
}


//////////////////////////////////////////////////////
//
//  PRIVATE
//
uint8_t TM1637::writeByte(uint8_t data)
{
  //  shift out data 8 bits LSB first
  for (uint8_t i = 8; i > 0; i--)
  {
    writeSync(_clockPin, LOW);
    writeSync(_dataPin, data & 0x01);
    writeSync(_clockPin, HIGH);
    data >>= 1;
  }

  writeSync(_clockPin, LOW);
  writeSync(_dataPin, HIGH);
  writeSync(_clockPin, HIGH);

  //  get ACKNOWLEDGE
  pinMode(_dataPin, INPUT);
  delayMicroseconds(_bitDelay);
  uint8_t rv = digitalRead(_dataPin);

  //  FORCE OUTPUT LOW
  pinMode(_dataPin, OUTPUT);
  digitalWrite(_dataPin, LOW);
  delayMicroseconds(_bitDelay);
  return rv;
}


void TM1637::start()
{
  writeSync(_clockPin, HIGH);
  writeSync(_dataPin, HIGH);
  writeSync(_dataPin, LOW);
  writeSync(_clockPin, LOW);
}


void TM1637::stop()
{
  writeSync(_clockPin, LOW);
  writeSync(_dataPin, LOW);
  writeSync(_clockPin, HIGH);
  writeSync(_dataPin, HIGH);
}


void TM1637::writeSync(uint8_t pin, uint8_t val)
{
  digitalWrite(pin, val);

  #if defined(ESP32)
    nanoDelay(21);  //  delay(2) is not enough in practice.
  #endif
  //  other processors may need other "nanoDelay(n)"
}


//  keyscan results are reversed left for right from the data sheet.
//  here are the values returned by keyscan():
//
//  pin         2     3     4     5     6     7     8     9
//            sg1   sg2   sg3   sg4   sg5   sg6   sg7   sg8
//  19   k1  0xf7  0xf6  0xf5  0xf4  0xf3  0xf2  0xf1  0xf0
//  20   k2  0xef  0xee  0xed  0xec  0xeb  0xea  0xe9  0xe8

uint8_t TM1637::keyscan(void)
{
  uint8_t halfDelay = _bitDelay >> 1;
  uint8_t key;
  start();
  key = 0;
  writeByte(TM1637_READ_KEYSCAN);  //  includes the ACK, leaves DATA low
  pinMode(_dataPin, INPUT_PULLUP);

  for (uint8_t i = 0; i <= 7; i++) {
    writeSync(_clockPin, LOW);
    delayMicroseconds(halfDelay);
    writeSync(_clockPin, HIGH);
    delayMicroseconds(halfDelay);
    key >>= 1;
    key |= (digitalRead(_dataPin)) ? 0x80 : 0x00 ;
  }

  writeSync(_clockPin, LOW);
  delayMicroseconds(halfDelay);
  writeSync(_clockPin, HIGH);

  //  wait for ACK
  delayMicroseconds(halfDelay);

  //  FORCE OUTPUT LOW
  pinMode(_dataPin, OUTPUT);
  digitalWrite(_dataPin, LOW);
  delayMicroseconds(halfDelay);
  stop();
  return key;
}


//  nanoDelay() makes it possible to go into the sub micron delays.
//  It is used to lengthen pulses to be minimal 400 ns but not much longer.
//  See datasheet.
void TM1637::nanoDelay(uint16_t n)
{
  volatile uint16_t i = n;
  while (i--);
}


uint8_t TM1637::asciiTo7Segment ( char c )
{
  /*
      -01-
  20 |    | 02
      -40-
  10 |    | 04
      -08-     .80
  */
  //  7+1  Segment patterns for ASCII 0x30-0x5F
  const uint8_t asciiToSegments[] = {
    0x3f,0x06,0x5b,0x4f, 0x66,0x6d,0x7d,0x07,  //  0123 4567
    0x7f,0x6f,0x09,0x89, 0x58,0x48,0x4c,0xD3,  //  89:; <=>?
    0x5f,0x77,0x7c,0x39, 0x5E,0x79,0x71,0x3d,  //  @ABC DEFG
    0x76,0x06,0x0E,0x75, 0x38,0x37,0x54,0x5c,  //  HIJK LMNO
    0x73,0x67,0x50,0x6D, 0x78,0x3E,0x1C,0x9c,  //  PQRS TUVW
    0x76,0x6E,0x5B,0x39, 0x52,0x0F,0x23,0x08   //  XYZ[ /]^_
  };

  uint8_t segments = c & 0x80;
  c &= 0x7f;
  if ( c >= 0x60 ) c -= 0x20 ;                 //  a-z -> A-Z
  if ( c == '.' ) segments = 0x80;             //  decimal point only
  if ( c == '-' ) segments |= 0x40;            //  minus sign
  if ( ( c >= 0x30 ) && ( c <= 0x5F ) ) {
    segments |= asciiToSegments[ c - 0x30 ];
  }
  return segments;
}


//  -- END OF FILE --

